src/wp-includes/speculative-loading.php

```diff
@@ -0,0 +1,268 @@
+		$config['mode'] = $default_mode;
+	}
+	if ( ! isset( $config['eagerness'] ) || 'auto' === $config['eagerness'] || ! wp_is_valid_speculation_rules_eagerness( $config['eagerness'] ) ) {
+		$config['eagerness'] = $default_eagerness;
+	}
+
+	return array(
+		'mode'      => $config['mode'],
+		'eagerness' => $config['eagerness'],
+	);
+}
+
+/**
+ * Checks whether the given speculation rules mode is valid.
+ *
+ * @since 6.8.0
+ *
+ * @param string $mode Speculation rules mode.
+ * @return bool True if valid, false otherwise.
+ */
+function wp_is_valid_speculation_rules_mode( string $mode ): bool {
+	static $mode_allowlist = array(
+		'prefetch'  => true,
+		'prerender' => true,
+	);
+	return isset( $mode_allowlist[ $mode ] );
+}
+
+/**
+ * Checks whether the given speculation rules eagerness is valid.
+ *
+ * @since 6.8.0
+ *
+ * @param string $eagerness Speculation rules eagerness.
+ * @return bool True if valid, false otherwise.
+ */
+function wp_is_valid_speculation_rules_eagerness( string $eagerness ): bool {
+	static $eagerness_allowlist = array(
+		'eager'        => true,
+		'moderate'     => true,
+		'conservative' => true,
+	);
+	return isset( $eagerness_allowlist[ $eagerness ] );
+}
+
+/**
+ * Returns the full speculation rules data based on the given configuration.
+ *
+ * Plugins with features that rely on frontend URLs to exclude from prefetching or prerendering should use the
+ * {@see 'wp_speculation_rules_href_exclude_paths'} filter to ensure those URL patterns are excluded.
+ *
+ * @since 6.8.0
+ * @access private
+ *
+ * @param array<string, string> $configuration Associative array with 'mode' and 'eagerness' keys.
+ * @return array<string, array<int, array<string, mixed>>> Associative array of speculation rules by type.
+ */
+function wp_get_speculation_rules( array $configuration ): array {
+	// If the passed configuration is invalid, trigger a warning and fall back to the default configuration.
+	if (
+		! isset( $configuration['mode'] ) ||
+		! wp_is_valid_speculation_rules_mode( $configuration['mode'] ) ||
+		! isset( $configuration['eagerness'] ) ||
+		! wp_is_valid_speculation_rules_eagerness( $configuration['eagerness'] )
+	) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			esc_html(
+				sprintf(
+					/* translators: %s is $configuration */
+					__( 'The %s parameter was provided with an invalid value.' ),
+					'$configuration'
+				)
+			),
+			'6.8.0'
+		);
+		$configuration = wp_get_speculation_rules_configuration();
+		if ( null === $configuration ) {
+			return array();
+		}
+	}
+
+	$mode      = $configuration['mode'];
+	$eagerness = $configuration['eagerness'];
+
+	$prefixer = new WP_URL_Pattern_Prefixer();
+
+	$base_href_exclude_paths = array(
+		$prefixer->prefix_path_pattern( '/wp-*.php', 'site' ),
+		$prefixer->prefix_path_pattern( '/wp-admin/*', 'site' ),
+		$prefixer->prefix_path_pattern( '/*', 'uploads' ),
```

Comment:
could we in most cases exclude `wp-content/*` instead of each of these individual paths (unless they were in non-standard locations without the wp-content prefix)?

Reply:
None of these three paths is actually within `wp-content`. Note that the `wp-content` directory is already excluded, see line 162.

Reply:
I suppose the uploads path is excluded separately because it's not necessarily under `wp-content`, right?

Reply:
Yeah, same with plugins and themes. 99% of times they should be within `wp-content`, but not guaranteed.

Reply:
True, but it also doesn't hurt.

Reply:
Agreed it doesn't hurt, checking for that would likely be more expensive to do than including the extra rules unconditionally.